File: src/main/java/seremis/geninfusion/soul/standardSoul/StandardSoulZombie.java
package seremis.geninfusion.soul.standardSoul;

import net.minecraft.client.model.ModelSkeleton;
import net.minecraft.client.model.ModelZombie;
import net.minecraft.entity.EntityLiving;
import net.minecraft.entity.passive.EntityVillager;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import seremis.geninfusion.api.soul.IChromosome;
import seremis.geninfusion.api.soul.lib.Genes;
import seremis.geninfusion.api.soul.util.ModelPart;
import seremis.geninfusion.soul.Chromosome;
import seremis.geninfusion.soul.allele.*;

public class StandardSoulZombie extends StandardSoul {

    @Override
    public IChromosome getChromosomeFromGene(EntityLiving entity, String gene) {
        if(gene.equals(Genes.GENE_BURNS_IN_DAYLIGHT))
            return new Chromosome(new AlleleBoolean(false, true));
        if(gene.equals(Genes.GENE_DEATH_SOUND))
            return new Chromosome(new AlleleString(true, "mob.zombie.death"));
        if(gene.equals(Genes.GENE_HURT_SOUND))
            return new Chromosome(new AlleleString(false, "mob.zombie.hurt"));
        if(gene.equals(Genes.GENE_ITEM_DROPS))
            return new Chromosome(new AlleleInventory(false, new ItemStack[]{new ItemStack(Items.rotten_flesh)}));
        if(gene.equals(Genes.GENE_LIVING_SOUND))
            return new Chromosome(new AlleleString(false, "mob.zombie.say"));
        if(gene.equals(Genes.GENE_PICKS_UP_ITEMS))
            return new Chromosome(new AlleleBoolean(false, true));
        if(gene.equals(Genes.GENE_RARE_ITEM_DROPS))
            return new Chromosome(new AlleleInventory(false, new ItemStack[]{new ItemStack(Items.iron_ingot), new ItemStack(Items.carrot), new ItemStack(Items.potato)}));
        if(gene.equals(Genes.GENE_SET_ON_FIRE_FROM_ATTACK))
            return new Chromosome(new AlleleBoolean(false, true));
        if(gene.equals(Genes.GENE_SPLASH_SOUND))
            return new Chromosome(new AlleleString(false, "game.hostile.swim.splash"));
        if(gene.equals(Genes.GENE_SWIM_SOUND))
            return new Chromosome(new AlleleString(false, "game.hostile.swim"));
        if(gene.equals(Genes.GENE_WALK_SOUND))
            return new Chromosome(new AlleleString(true, "mob.zombie.step"));
        if(gene.equals(Genes.GENE_USE_NEW_AI))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_USE_OLD_AI))
            return new Chromosome(new AlleleBoolean(true, false));

        if(gene.equals(Genes.GENE_AI_SWIMMING))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_SWIMMING_INDEX))
            return new Chromosome(new AlleleInteger(true, 0));

        if(gene.equals(Genes.GENE_AI_ATTACK_ON_COLLIDE))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_ATTACK_ON_COLLIDE_INDEX))
            return new Chromosome(new AlleleIntArray(true, new int[]{2, 4}));
        if(gene.equals(Genes.GENE_AI_ATTACK_ON_COLLIDE_LONG_MEMORY))
            return new Chromosome(new AlleleBooleanArray(true, new boolean[]{false, true}));
        if(gene.equals(Genes.GENE_AI_ATTACK_ON_COLLIDE_MOVE_SPEED))
            return new Chromosome(new AlleleDoubleArray(true, new double[]{1.0D, 1.0D}));
        if(gene.equals(Genes.GENE_AI_ATTACK_ON_COLLIDE_TARGET))
            return new Chromosome(new AlleleClassArray(true, new Class[]{EntityPlayer.class, EntityVillager.class}));

        if(gene.equals(Genes.GENE_AI_MOVE_TOWARDS_RESTRICTION))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_MOVE_TOWARDS_RESTRICTION_INDEX))
            return new Chromosome(new AlleleInteger(true, 5));
        if(gene.equals(Genes.GENE_AI_MOVE_TOWARDS_RESTRICTION_MOVE_SPEED))
            return new Chromosome(new AlleleDouble(false, 1.0D));

        if(gene.equals(Genes.GENE_AI_MOVE_THROUGH_VILLAGE))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_MOVE_THROUGH_VILLAGE_INDEX))
            return new Chromosome(new AlleleInteger(true, 6));
        if(gene.equals(Genes.GENE_AI_MOVE_THROUGH_VILLAGE_IS_NOCTURNAL))
            return new Chromosome(new AlleleBoolean(false, true));
        if(gene.equals(Genes.GENE_AI_MOVE_THROUGH_VILLAGE_MOVE_SPEED))
            return new Chromosome(new AlleleDouble(true, 1.0D));

        if(gene.equals(Genes.GENE_AI_WANDER))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_WANDER_INDEX))
            return new Chromosome(new AlleleInteger(true, 7));
        if(gene.equals(Genes.GENE_AI_WANDER_MOVE_SPEED))
            return new Chromosome(new AlleleDouble(true, 1.0D));

        if(gene.equals(Genes.GENE_AI_WATCH_CLOSEST))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_WATCH_CLOSEST_INDEX))
            return new Chromosome(new AlleleIntArray(true, new int[]{8}));
        if(gene.equals(Genes.GENE_AI_WATCH_CLOSEST_CHANCE))
            return new Chromosome(new AlleleFloatArray(true, new float[]{0.02F}));
        if(gene.equals(Genes.GENE_AI_WATCH_CLOSEST_RANGE))
            return new Chromosome(new AlleleFloatArray(true, new float[]{8.0F}));
        if(gene.equals(Genes.GENE_AI_WATCH_CLOSEST_TARGET))
            return new Chromosome(new AlleleClassArray(true, new Class[]{EntityPlayer.class}));

        if(gene.equals(Genes.GENE_AI_LOOK_IDLE))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_LOOK_IDLE_INDEX))
            return new Chromosome(new AlleleInteger(true, 8));

        if(gene.equals(Genes.GENE_AI_HURT_BY_TARGET))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_HURT_BY_TARGET_INDEX))
            return new Chromosome(new AlleleInteger(true, 1));
        if(gene.equals(Genes.GENE_AI_HURT_BY_TARGET_CALL_HELP))
            return new Chromosome(new AlleleBoolean(true, true));

        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET))
            return new Chromosome(new AlleleBoolean(true, true));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_INDEX))
            return new Chromosome(new AlleleIntArray(true, new int[]{2, 4}));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_ENTITY_SELECTOR))
            return new Chromosome(new AlleleStringArray(true, new String[]{"", ""}));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_NEARBY_ONLY))
            return new Chromosome(new AlleleBooleanArray(true, new boolean[]{false, false}));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_TARGET))
            return new Chromosome(new AlleleClassArray(false, new Class[]{EntityPlayer.class, EntityVillager.class}));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_TARGET_CHANCE))
            return new Chromosome(new AlleleIntArray(true, new int[]{0, 0}));
        if(gene.equals(Genes.GENE_AI_NEAREST_ATTACKABLE_TARGET_VISIBLE))
            return new Chromosome(new AlleleBooleanArray(true, new boolean[]{true, false}));

        if(gene.equals(Genes.GENE_MODEL))
            return new Chromosome(new AlleleModelPartArray(true, ModelPart.getModelPartsFromModel(new ModelZombie(), entity)));
        if(gene.equals(Genes.GENE_TEXTURE))
            return new Chromosome(new AlleleString(true, "textures/entity/zombie/zombie.png"), new AlleleString(false, "textures/entity/zombie/zombie.png"));


        return super.getChromosomeFromGene(entity, gene);
    }
}

File: src/main/scala/seremis/geninfusion/soul/standardSoul/StandardSoulZombie.scala
package seremis.geninfusion.soul.standardSoul

import net.minecraft.client.model.ModelZombie
import net.minecraft.entity.EntityLiving
import net.minecraft.entity.monster.EntityZombie
import net.minecraft.entity.passive.EntityVillager
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import seremis.geninfusion.api.lib.Genes._
import seremis.geninfusion.api.soul.IChromosome
import seremis.geninfusion.soul.{Allele, Chromosome}

class StandardSoulZombie extends StandardSoul {

    override def getStandardSoulEntity: Class[_ <: EntityLiving] = classOf[EntityZombie]

    override def getChromosomeFromGene(entity: EntityLiving, gene: String): IChromosome = {
        if(gene.equals(GeneBurnsInDaylight))
            return new Chromosome(gene, new Allele(false, true, classOf[Boolean]))
        if(gene.equals(GeneDeathSound))
            return new Chromosome(gene, new Allele(true, "mob.zombie.death", classOf[String]))
        if(gene.equals(GeneHurtSound))
            return new Chromosome(gene, new Allele(false, "mob.zombie.hurt", classOf[String]))
        if(gene.equals(GeneItemDrops))
            return new Chromosome(gene, new Allele(false, Array(new ItemStack(Items.rotten_flesh)), classOf[Array[ItemStack]]))
        if(gene.equals(GeneLivingSound))
            return new Chromosome(gene, new Allele(false, "mob.zombie.say", classOf[String]))
        if(gene.equals(GenePicksUpItems))
            return new Chromosome(gene, new Allele(false, true, classOf[Boolean]))
        if(gene.equals(GeneRareItemDrops))
            return new Chromosome(gene, new Allele(false, Array(new ItemStack(Items.iron_ingot), new ItemStack(Items.carrot), new ItemStack(Items.potato)), classOf[Array[ItemStack]]))
        if(gene.equals(GeneSetOnFireFromAttack))
            return new Chromosome(gene, new Allele(false, true, classOf[Boolean]))
        if(gene.equals(GeneSplashSound))
            return new Chromosome(gene, new Allele(false, "game.hostile.swim.splash", classOf[String]))
        if(gene.equals(GeneSwimSound))
            return new Chromosome(gene, new Allele(false, "game.hostile.swim", classOf[String]))
        if(gene.equals(GeneWalkSound))
            return new Chromosome(gene, new Allele(true, "mob.zombie.step", classOf[String]))
        if(gene.equals(GeneUseNewAI))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneUseOldAI))
            return new Chromosome(gene, new Allele(true, false, classOf[Boolean]))

        if(gene == GeneChildSpeedModifier)
            return new Chromosome(gene, new Allele(false, 0.5D, classOf[Double]))
        if(gene == GeneChildXPModifier)
            return new Chromosome(gene, new Allele(true, 2.5F, classOf[Float]))

        if(gene.equals(GeneAISwimming))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAISwimmingIndex))
            return new Chromosome(gene, new Allele(true, 0, classOf[Int]))

        if(gene.equals(GeneAIAttackOnCollide))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIAttackOnCollideIndex))
            return new Chromosome(gene, new Allele(true, Array(2, 4), classOf[Array[Int]]))
        if(gene.equals(GeneAIAttackOnCollideLongMemory))
            return new Chromosome(gene, new Allele(true, Array(false, true), classOf[Array[Boolean]]))
        if(gene.equals(GeneAIAttackOnCollideMoveSpeed))
            return new Chromosome(gene, new Allele(true, Array(1.0D, 1.0D), classOf[Array[Double]]))
        if(gene.equals(GeneAIAttackOnCollideTarget))
            return new Chromosome(gene, new Allele(true, Array(classOf[EntityPlayer], classOf[EntityVillager]), classOf[Array[Class[_]]]))

        if(gene.equals(GeneAIMoveTowardsRestriction))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIMoveTowardsRestrictionIndex))
            return new Chromosome(gene, new Allele(true, 5, classOf[Int]))
        if(gene.equals(GeneAIMoveTowardsRestrictionMoveSpeed))
            return new Chromosome(gene, new Allele(false, 1.0D, classOf[Double]))

        if(gene.equals(GeneAIMoveThroughVillage))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIMoveThroughVillageIndex))
            return new Chromosome(gene, new Allele(true, 6, classOf[Int]))
        if(gene.equals(GeneAIMoveThroughVillageIsNocturnal))
            return new Chromosome(gene, new Allele(false, true, classOf[Boolean]))
        if(gene.equals(GeneAIMoveThroughVillageMoveSpeed))
            return new Chromosome(gene, new Allele(true, 1.0D, classOf[Double]))

        if(gene.equals(GeneAIWander))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIWanderIndex))
            return new Chromosome(gene, new Allele(true, 7, classOf[Int]))
        if(gene.equals(GeneAIWanderMoveSpeed))
            return new Chromosome(gene, new Allele(true, 1.0D, classOf[Double]))

        if(gene.equals(GeneAIWatchClosest))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIWatchClosestIndex))
            return new Chromosome(gene, new Allele(true, Array(8), classOf[Array[Int]]))
        if(gene.equals(GeneAIWatchClosestChance))
            return new Chromosome(gene, new Allele(true, Array(0.02F), classOf[Array[Float]]))
        if(gene.equals(GeneAIWatchClosestRange))
            return new Chromosome(gene, new Allele(true, Array(8.0F), classOf[Array[Float]]))
        if(gene.equals(GeneAIWatchClosestTarget))
            return new Chromosome(gene, new Allele(true, Array(classOf[EntityPlayer]), classOf[Array[Class[_]]]))

        if(gene.equals(GeneAILookIdle))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAILookIdleIndex))
            return new Chromosome(gene, new Allele(true, 8, classOf[Int]))

        if(gene.equals(GeneAIHurtByTarget))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAIHurtByTargetIndex))
            return new Chromosome(gene, new Allele(true, 1, classOf[Int]))
        if(gene.equals(GeneAIHurtByTargetCallHelp))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))

        if(gene.equals(GeneAINearestAttackableTarget))
            return new Chromosome(gene, new Allele(true, true, classOf[Boolean]))
        if(gene.equals(GeneAINearestAttackableTargetIndex))
            return new Chromosome(gene, new Allele(true, Array(2, 4), classOf[Array[Int]]))
        if(gene.equals(GeneAINearestAttackableTargetEntitySelector))
            return new Chromosome(gene, new Allele(true, Array("", ""), classOf[Array[String]]))
        if(gene.equals(GeneAINearestAttackableTargetNearbyOnly))
            return new Chromosome(gene, new Allele(true, Array(false, false), classOf[Array[Boolean]]))
        if(gene.equals(GeneAINearestAttackableTargetTarget))
            return new Chromosome(gene, new Allele(false, Array(classOf[EntityPlayer], classOf[EntityVillager]), classOf[Array[Class[_]]]))
        if(gene.equals(GeneAINearestAttackableTargetTargetChance))
            return new Chromosome(gene, new Allele(true, Array(0, 0), classOf[Array[Int]]))
        if(gene.equals(GeneAINearestAttackableTargetVisible))
            return new Chromosome(gene, new Allele(true, Array(true, false), classOf[Array[Boolean]]))

        if(gene.equals(GeneModelAdult))
            return new Chromosome(gene, new Allele(true, model, classOf[Model]))
        if(gene.equals(GeneTexture))
            return new Chromosome(gene, new Allele(true, textureStringToNBT("textures/entity/zombie/zombie.png"), classOf[NBTTagCompound]), new Allele(false, textureStringToNBT("textures/entity/zombie/zombie.png"), classOf[NBTTagCompound]))


        super.getChromosomeFromGene(entity, gene)
    }

    val model: Model = {
        val modelZombie = new ModelZombie()

        modelZombie.bipedRightArm.rotateAngleX = -1.57F
        modelZombie.bipedLeftArm.rotateAngleX = -1.57F

        modelBiped(modelZombie, true)
    }
}
